### DISCUSSION TITLE
HOWTO Compiling under windows with windows clang

### DISCUSSION THREAD

**Discussion by nick-l-o3de** [o3de/o3de#13710] [Developer Articles]
as of commit [38c3a9d81a4edb9c6260c3a17165b634c2e1c191 (](https://github.com/o3de/o3de/commit/38c3a9d81a4edb9c6260c3a17165b634c2e1c191) it is possible to compile under llvm-clang on windows.   This is a brief guide how.

You will need:  
* python 3.10+ on your path.  You can use the one from o3de, but you don't have to.  To use the one from O3DE, `set PATH=%PATH%;C:\Users\YOUR_USER_NAME\.o3de\Python\packages\python-3.10.13-rev1-windows\python`
* cmake 3.24+ on your path (`winget install Kitware.CMake`)
* ninja build installed and on your path. (`winget install  Ninja-build.Ninja`)
* The windows ATL libraries (only required for lldb the debugger you can omit this if you dont want the debugger)
* the appropriate vcvarsall.bat activated on your path (see below)

VCVARS: 
clang (every version) will use whatever Microsoft C-Runtime is active from your vcvarsall.bat and the Microsoft C-Runtime is sensitive to what version of clang you're using - for example, VS2022 CRT (v14.44) will static assert if you use clang older than 19.

So for example, if you want to use clang14-18 you must also use the VS2019 CRT (v14.29).  You can see what runtimes are installed here: `C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC` and you can install more runtimes with the Visual Studio Installer.  VS2022 can use the VS2019 runtimes - you don't have to downgrade visual studio itself.

in visual studio installer, you can install the compile tools and atl for searching for `C++/CLI support for v143 build tools (latest)` and installing the same (v143) ATL as shows up in your `C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC` runtimes folder, eg, if you see a 14.29 folder there, look for the 14.29 ATL in the msvc installer.

To activate the runtime in a command prompt add `-vcvars_ver` to the end of the invocation, ie
```cmd
for vs2019:
"C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Auxiliary\Build\vcvarsall.bat" amd64 -vcvars_ver=14.29
or for vs2022:
"C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Auxiliary\Build\vcvarsall.bat" amd64 -vcvars_ver=14.44
```

If all you want to do is build O3DE with a pre-built windows version of clang, you can get a pre-built clang binary from, say https://github.com/llvm/llvm-project/releases and unzip it and ensure that the bin folder is in your path) ie, clang++ must be found when you type clang++ in a cmd window.

On the other hand if you want to compile or use tools like https://github.com/include-what-you-use/include-what-you-use you will need to build clang yourself or find a binary that includes its libraries, headers, etc.   This will briefly cover doing that, since its much more interesting to be able to use those tools than it is to just compile O3DE using clang on windows.

### Compile clang on windows

In this example, I'll be doing clang15, so I must use vcvarsver `14.29` and `-Tv142,host=x64`  to build with vs2019 crt
If I was using clang19 or above, I could use vcvars `14.44` to build with 2022 and I'd specify `-Tv143,host=x64`

```
"C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Auxiliary\Build\vcvarsall.bat" amd64 -vcvars_ver=14.29
cd\
mkdir playground
cd playground
git clone https://github.com/llvm/llvm-project.git
cd llvm-project
git checkout release/15.x
cmake -B build -S llvm -DLLVM_ENABLE_PROJECTS=clang;lld;lldb -A x64 -Tv142,host=x64  -DCMAKE_INSTALL_PREFIX=../llvm_install -DLLVM_TARGETS_TO_BUILD="X86"
cmake --build build --config Release --target INSTALL --parallel
```
Explanation
`LLVM_ENABLE_PROJECTS` - we want the clang compiler, the clang linker `lld` and the clang debugger `lldb`.
`CMAKE_INSTALL_PREFIX` - where to put the completed binaries
`LLVM_TARGETS_TO_BUILD` - we only care about x86/x64.  This is where you'd tell it to include ARM/ARM64, MIPS, and other architectures.
`-Tv142,Host=x64` - we are building using vs2019 aka v142, and we want to target x64 binaries.  a 32 bit compiler will run out of memory

This will build and install the llvm/clang tools to a folder called llvm_install in the playground folder.

You now have a working llvm/clang.  you can now use that to compile o3de or other tools which depend on it.

### compile o3de with clang
add the llvm/clang tools (which you downloaded or built) into your path, ie
```
set PATH=(FOLDER_TO_llvm_install/bin subfolder);%PATH%
```
at that point you should be able to type "clang++ --version" and see the right one.
Go to your o3de root folder or your project's folder or whatever you want to build.  (If you're in a project folder, do not supply LY_PROJECTS)
```
cmake -S . -B build\clang_windows -DLY_PROJECTS=AutomatedTesting -G "Ninja Multi-Config" -DCMAKE_C_COMPILER="d:/llvm/llvm_install/bin/clang.exe" -DCMAKE_CXX_COMPILER="d:/playground/llvm_install/bin\clang++.exe"  -DCMAKE_RC_COMPILER="d:/llvm/llvm_install_14/bin/llvm-rc.exe"

cmake --build build/clang_windows --config profile --target Editor --parallel
```

The binaries will be in (o3de)/build/clang_windows/bin/profile.  You can choose a different target (like AssetProcessor)

### compile include-what-you-use with your custom clang
You need to clone and compile it and let it know where you put your custom clang.
I'm assuming you wanna check it out and build it in the same playground folder you built llvm into but feel free to modify this:
```
set PATH=(path-to-your-llvm_install/bin-folder);%PATH%
cd\playground
git clone https://github.com/include-what-you-use/include-what-you-use.git
cd include-what-you-use
git checkout clang_15
cmake -G "Ninja Multi-Config" -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_PREFIX_PATH=D:\playground\llvm_install -S . -B build
cmake --build build --config Release --parallel
```

This builds include-what-you-use into playground\include-what-you-use\build\bin\Release

You can then hook it into o3de like so, in your o3de folder.
```
cmake -S . -B build\clang_windows -DCMAKE_C_COMPILER=clang  -DCMAKE_CXX_COMPILER=clang++ -DLY_PROJECTS=AutomatedTesting -G "Ninja Multi-Config" -DLY_UNITY_BUILD=0 -DCMAKE_CXX_INCLUDE_WHAT_YOU_USE="d:\playground\include-what-you-use\build\bin\Release\include-what-you-use.exe;-Xiwyu;--cxx17ns;-Xiwyu;--comment_style=short;-Xiwyu;--verbose=3" 
cmake --build build/clang_windows --config profile
```
replacing d:\playground path to where you built it of course, and verbose level with whatever you want.

### Clang Build Analyzer
Can generate time-trace information that tells you what the most costly functions to compile, costly headers, etc.
Similar to above, assume you're already set up with llvm/clang and have already added its binfolder to your path so that clang++ works.
```
cd\playground
git clone https://github.com/aras-p/ClangBuildAnalyzer.git
cd ClangBuildAnalyzer
cmake -B build -S . -DCMAKE_INSTALL_PREFIX=install
cmake --build build --config Release --target INSTALL
copy ClangBuildAnalyzer.ini install\bin
```
You can now run it from (playground)\ClangBuildAnalyzer\install\bin

#### Using ClangBuildAnalyzer in o3de
in the o3de folder, assuming you already have all the above set(clang++ on path, etc)

```
cmake -S . -B -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ build/clang_windows -DLY_PROJECTS=AutomatedTesting -G "Ninja Multi-Config" -DO3DE_EXTRA_CXX_FLAGS=-ftime-trace -DO3DE_EXTRA_C_FLAGS=-ftime-trace -DLY_UNITY_BUILD=0 -UCMAKE_CXX_INCLUDE_WHAT_YOU_USE
```
note we disable include what you use before we try to use clang analyzer.  Do not recommend using both at same time.
We also add O3DE_EXTRA_C_FLAGS and CXX flags to be -ftime-trace.  This adds that param to every single compile.

ClangBuildAnalyzer requires you to start it before you build, then build, then finish it for analysis after.  There are other ways to run it, but here's an example.  (Assumes you installed Clang Build Analyzer in d:\playground\ClangBuildAnalyzer.  Change it to match what you did.

```
d:\playground\ClangBuildAnalyzer\install\bin\ClangBuildAnalyzer.exe --start build\clang_windows

cmake --build build/clang_windows --config Profile --target AzCore

d:\playground\ClangBuildAnalyzer\install\bin\ClangBuildAnalyzer.exe --stop build\clang_windows outputfile.dat
d:\playground\ClangBuildAnalyzer\install\bin\ClangBuildAnalyzer.exe --analyze outputfile.dat > analysis.txt
```

This outputs analysis.txt full of useful information.  -ftime-trace also outputs a profiler time trace of the actual compilation of each compiel unit, next to the object file.  You can dig in the build folder for it an upload it to a tool like speedscope to view it.

**Comment by Jackie9527** [comment]
With the following commands, we can build o3de with clang-cl shipped with VS2022.
```
cmake -S . ^
    -B build/clang_windows ^
    -T ClangCl ^
    -DCMAKE_C_COMPILER=clang ^
    -DCMAKE_CXX_COMPILER=clang++ ^
    -DLY_UNITY_BUILD=ON ^
    -DLY_PROJECTS=AutomatedTesting

cmake --build build/clang_windows --config profile
```

The version of clang shipped with VS2022, as follows:
```
clang --version
clang version 15.0.1
Target: x86_64-pc-windows-msvc
Thread model: posix
```

Known issues:
#14943 
#14942 
[186](https://github.com/o3de/3p-package-source/pull/186)

A PR  is underway [15018](https://github.com/o3de/o3de/pull/15018).

**Comment by mythrz** [comment]
Doing some tests without Visual Studio or its tools! Installed clang and CMake manually. But it is failing!

Can any one confirm or not if the engine can still compile under clang? Should I try different flags? Change the CMake files?

Tested with: 

O3DE commit: f384a0ad8ba373f15a81fce1322ea16b888c6f98

```
clang version 19.1.0
Target: x86_64-pc-windows-msvc
Thread model: posix
InstalledDir: C:\Program Files\LLVM\bin
```

`cmake version 3.31.0`

```
cmake -S . -B build/clang_windows -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++  -DLY_PROJECTS=AutomatedTesting -G "Ninja Multi-Config"
cmake --build build/clang_windows --config profile
```
or
`cmake --build build/clang_windows --config profile --target Editor`

Not able to compile.

from 16/2597 forward all fail. It stops at 18/2597 e.g.,

`
[16/2597] Building CXX object Code/Framework/AzCore/CMakeFiles/O3DEKernel.dir/profile/Unity/unity_0_cxx.cxx.obj
FAILED: Code/Framework/AzCore/CMakeFiles/O3DEKernel.dir/profile/Unity/unity_0_cxx.cxx.obj
C:\PROGRA~1\LLVM\bin\CLANG_~1.EXE -DAZ_BUILD_CONFIGURATION_TYPE=\"profile\" -DAZ_ENABLE_DEBUG_TOOLS -DAZ_ENABLE_TRACING -DAZ_PROFILE_BUILD -DNDEBUG -DNOMINMAX -DO3DEKernel_EXPORTS -DWIN32 -DWIN64 -D_ENABLE_EXTENDED_ALIGNED_STORAGE -D_FORTIFY_SOURCE=2 -D_HAS_EXCEPTIONS=0 -D_PROFILE -D_WIN32 -D_WIN64 -DCMAKE_INTDIR=\"profile\" -IC:/O3DE_ClangAndNinja/o3de/Code/Framework/AzCore/. -IC:/O3DE_ClangAndNinja/o3de/Code/Framework/AzCore/Platform/Windows -IC:/O3DE_ClangAndNinja/o3de/Code/Framework/AzCore/Platform/Common -fno-exceptions -fvisibility=hidden -fvisibility-inlines-hidden -Wall -Werror -Wno-inconsistent-missing-override -Wrange-loop-analysis -Wno-unknown-warning-option -Wno-parentheses -Wno-reorder -Wno-switch -Wno-undefined-var-template -msse3 -mf16c -Wno-deprecated-declarations  -O2 -g -fstack-protector-all -fstack-check -std=c++17 -D_DLL -D_MT -Xclang --dependent-lib=msvcrt -MD -MT Code/Framework/AzCore/CMakeFiles/O3DEKernel.dir/profile/Unity/unity_0_cxx.cxx.obj -MF Code\Framework\AzCore\CMakeFiles\O3DEKernel.dir\profile\Unity\unity_0_cxx.cxx.obj.d -o Code/Framework/AzCore/CMakeFiles/O3DEKernel.dir/profile/Unity/unity_0_cxx.cxx.obj -c C:/O3DE_ClangAndNinja/o3de/build/clang_windows/Code/Framework/AzCore/CMakeFiles/O3DEKernel.dir/Unity/unity_0_cxx.cxx
In file included from C:/O3DE_ClangAndNinja/o3de/build/clang_windows/Code/Framework/AzCore/CMakeFiles/O3DEKernel.dir/Unity/unity_0_cxx.cxx:4:
In file included from C:/O3DE_ClangAndNinja/o3de/Code/Framework/AzCore/AzCore/Module/Environment.cpp:11:
In file included from C:/O3DE_ClangAndNinja/o3de/Code/Framework/AzCore/.\AzCore/Memory/OSAllocator.h:11:
In file included from C:/O3DE_ClangAndNinja/o3de/Code/Framework/AzCore/.\AzCore/Memory/AllocatorBase.h:10:
In file included from C:/O3DE_ClangAndNinja/o3de/Code/Framework/AzCore/.\AzCore/Memory/IAllocator.h:11:
In file included from C:/O3DE_ClangAndNinja/o3de/Code/Framework/AzCore/.\AzCore/RTTI/RTTI.h:12:
C:/O3DE_ClangAndNinja/o3de/Code/Framework/AzCore/.\AzCore/RTTI/TypeInfo.h:518:5: error: call to 'GetTemplateIdentity' is ambiguous
  518 |     AZ_TYPE_INFO_INTERNAL_SPECIALIZED_TEMPLATE_BOTHFIX_UUID_DECL(AZStd::basic_fixed_string, AZ_TYPE_INFO_INTERNAL_TYPENAME, AZ_TYPE_INFO_INTERNAL_AUTO, AZ_TYPE_INFO_INTERNAL_TYPENAME);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
C:/O3DE_ClangAndNinja/o3de/Code/Framework/AzCore/.\AzCore/RTTI/TemplateInfo.h:277:5: note: expanded from macro 'AZ_TYPE_INFO_INTERNAL_SPECIALIZED_TEMPLATE_BOTHFIX_UUID_DECL'
  277 |     AZ_TEMPLATE_INFO_INTERNAL_BOTHFIX_UUID_DECL(_TemplateName) \
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
C:/O3DE_ClangAndNinja/o3de/Code/Framework/AzCore/.\AzCore/RTTI/TemplateInfo.h:263:18: note: expanded from macro 'AZ_TEMPLATE_INFO_INTERNAL_BOTHFIX_UUID_DECL'
  263 |         decltype(AZ::AzGenericTypeInfo::Internal::GetTemplateIdentity<_TemplateName>()));
      |                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
C:/O3DE_ClangAndNinja/o3de/Code/Framework/AzCore/.\AzCore/RTTI/TemplateInfo.h:108:10: note: candidate function [with T = AZStd::basic_fixed_string]
  108 |     auto GetTemplateIdentity() ->
      |          ^
C:/O3DE_ClangAndNinja/o3de/Code/Framework/AzCore/.\AzCore/RTTI/TemplateInfo.h:111:10: note: candidate function [with T = AZStd::basic_fixed_string]
  111 |     auto GetTemplateIdentity() -> TemplateIdentityTypeAutoType<T>;
      |          ^
1 error generated.
`

**Comment by nick-l-o3de** [comment]
I think that has been fixed in dev latest.

**Comment by kiotk** [comment]
I recently attempted to build O3DE using Clang (LLVM 19.1.7) on Windows, but after a successful build, launching the editor (Editor.exe) results in an application error (0xc0000142), preventing the editor from starting.

https://github.com/llvm/llvm-project/releases/download/llvmorg-19.1.7/LLVM-19.1.7-win64.exe

```
Installed Components in Visual Studio Build Tools 2022:

MSVC v143 - VS2022 C++ x64/x86 build tools(Latest)
C++ AddressSanitizer
Testing tools core features - Build Tools
C++ Build Tools core features 
C++ ATL for latest v143 build tools (x86 & x64)
C++ MFC for latest v143 build tools (x86 & x64)
Windows 10 SDK (10.0.19041.0)
Windows Universal C Runtime
MSBuild support for LLVM (clang-cl) toolset
C++ 2022 Redistributable Update
vcpkg package manager
Text Template Transformation
.NET SDK
.NET Framework 4.8 SDK
.NET Framework 4.8 targeting pack
.NET 9.0 Runtime
```

```
C:\o3de>cmake -S . -B build/windows -G "Ninja Multi-Config" 
-DLY_3RDPARTY_PATH=C:\Users\user\.o3de\3rdParty
-DLY_PROJECTS=C:/o3de/user/myproject 
-DLY_DISABLE_TEST_MODULES=ON 
-DCMAKE_C_COMPILER=clang 
-DCMAKE_CXX_COMPILER=clang++ 
```

`C:\o3de>cmake --build build/windows/ --config release --target Editor`

`C:\o3de>.\build\windows\bin\release\Editor.exe --project-path c:\o3de\user\myproject `

**ERROR: Editor.exe Application Error
The application was unable to start correctly (0xc0000142)
Click OK to close the application.**

**Comment by nick-l-o3de** [comment]
Yeah, I'm busy compiling and testing right now.  I was unable to start editor either, but I noticed it was because AssetProcessor crashes, and I noticed that assetprocessor crahses because assetbuilder fails to connect to it.  I'm thinking some sort of firewall or other socket issue. I'm building with lldb now (and have updated the build instructions)

Kinda unclear to me how come the software would work under msvc, and clang15 on linux, but not clang15 on windows. ( With no source changes between clang15 on linux and clang15 on windows) or why that would affect socket connectivity.  The difference is not in the code but the MSVCRT c runtime library, but .. why would the same code compiled in vc 2019 also work just fine (same crt)

I'm also testing clang18 and so far, although there's a bug with conditional explicit I will have to fix (PR incoming), it does start asset processor and assetbuilder just fine, I can open editor, etc So something between clang15 and clang18 causes socket issues somehow?

**Comment by nick-l-o3de** [comment]
I found that Clang18, Clang19 both work flawlessly for me on the stabilization branch as long as I patch the pair/tuple thing.  
Clang18 on windows cannot compile with out that fix.
Clang19 does not need that fix.

If somethings crashing we're going to need to figure out what makes it crash, specifically.  It doesn't happen to everyone, clearly

**Comment by nick-l-o3de** [comment]
https://github.com/o3de/o3de/pull/19247 fixes teh tuple and pair spam for clang18.

With this patch, on windows, at least, using the steps above to obtain each version of clang (buildingit) I was able to get the editor up an running on every clang except 15 and 14, due to the wierd assetbuilder thing, despite it succesfully compiling.

I'm going to switch to linux and try the same test just to get an idea of how it might work with a different C library.  On windows ,the only one worth using at the moment is the MS CRT so its not quite apples to apples to say "it compiles on every clang" if all I tested on is windows.

**Reply by nick-l-o3de** [reply]
I was only able to test on clang20 on linux, if I tried to use a different clang, it has all sorts of troubles - but, its my machine doing it, for exmaple, jenkins CI compiles it just fine on ubuntu/22.04/clang14 without issue and my machine running ubuntu 25.x cannot.  I will have to investigate containerization to try to get an apples to apples env.
